File: problems/sentimental/credit/credit.py
# Imports get_string and get_int function from cs50 library
from cs50 import get_string, get_int

# Do while python equivalent
while True:
    number = get_int("Number: ")
    if number > 0:
        break
# Converts integer to string
strnumber = str(number)
length = len(strnumber)
sum = 0
# Multiplies numbers and adds products' digits starting with number's second-to-last digit
for i in range(length - 2, -1, -2):
    mult = int(strnumber[i]) * 2
    ld = mult // 10
    rd = mult % 10
    sum += ld + rd
# Adds the remaining digits
for i in range(length - 1, -1, -2):
    sum += int(strnumber[i])
# Prints output
if sum % 10 == 0:
    td = int(strnumber[:2])
    if td == 34 or td == 37:
        print("AMEX")
    elif td in range(51, 56):
        print("MASTERCARD")
    elif int(strnumber[:1]) == 4:
        print("VISA")
else:
    print("INVALID")

File: problems/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

int main()
{
    string numstr = get_string("Number: ");
    int len = strlen(numstr);
    int sum = 0;
    // Convert string to long
    long long int num = atol(numstr);
    long long int tmp;
    long long int ten = 10;
    // Calculate products' digits
    do
    {
        tmp = num / ten;
        int dig = tmp % 10;
        int digs = dig * 2;
        int a = digs / 10;
        int b = digs % 10;
        sum += a + b;
        ten = ten * 100;
    }
    while (tmp != 0);
    ten = 1;
    tmp = num;
    // Sum the rest of numbers' digits
    do
    {
        int dig = tmp % 10;
        sum += dig;
        ten = ten * 100;
        tmp = num / ten;
    }
    while (tmp != 0);
    // Prints outputs
    if (sum % 10 == 0)
    {
        if (len == 13 || len == 16)
        {
            long long int pwr = pow(10, (len - 1));
            long long int pwr2 = pow(10, (len - 2));
            if (num / pwr == 4) // Check the first digit
            {
                printf("VISA\n");
            }
            else if (num / pwr2 >= 51 && num / pwr2 <= 55) // Check the first two digits
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (len == 15)
        {
            long long int pwr = pow(10, (len - 2));
            if (num / pwr == 34 || num / pwr == 37) // Check the first two digits
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n"); // Prints if numbers are invalid
    }
}